docs/ql-documentation/learn-ql/cpp/introduce-libraries-cpp.rst

```diff
@@ -4,114 +4,516 @@ Introducing the C/C++ libraries
+| **union U** ``{`` `Type <https://help.semmle.com/qldoc/cpp/semmle/code/cpp/Type.qll/type.Type$Type.html>`__ **variant1** ``;`` `Type <https://help.semmle.com/qldoc/cpp/semmle/code/cpp/Type.qll/type.Type$Type.html>`__ **variant2** ``,`` ... ``};``                                                                                                                                                                                      | `Union <hhttps://help.semmle.com/qldoc/cpp/semmle/code/cpp/Union.qll/type.Union$Union.html>`__                                                                        |                                                                                                                                                                                                                                                                                   |
++---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| **struct S** ``{`` `Type <https://help.semmle.com/qldoc/cpp/semmle/code/cpp/Type.qll/type.Type$Type.html>`__ **member_variable** ``;`` `Type <https://help.semmle.com/qldoc/cpp/semmle/code/cpp/Type.qll/type.Type$Type.html>`__ **member_function** ``(`` `Parameter <https://help.semmle.com/qldoc/cpp/semmle/code/cpp/Parameter.qll/type.Parameter$Parameter.html>`__ ... ``) {`` ... ``}``... ``};``                                    | `Struct <https://help.semmle.com/qldoc/cpp/semmle/code/cpp/Struct.qll/type.Struct$Struct.html>`__                                                                     |                                                                                                                                                                                                                                                                                   |
++-------------------------------------------------------------------------------------------------------------------------------------
```

Comment:
`Handler` is a bit hard to explain. As its documentation says, it has no surface syntax. It's the element _above_ each catch block (not below, so you need to change the `CatchBlock` and `CatchAnyBlock` examples).

In the control-flow graph, the `Handler` of `catch(const MyExn &e)` represents the point where there's a dynamic type test to decide whether the current exception is an instance of `MyExn`. There are two control-flow edges out of a handler: one that goes to the `CatchBlock` and one that goes either to the next `Handler` or out of the function.

I'm not sure what to pick for `Handler`'s syntactic example. I suggest "`catch` CatchBlock" to clarify how the two classes stand relative to each other in the AST. Then I'd also put `Handler` instead of `CatchBlock` for the two kinds of `TryStmt`.

Reply:
Since `Handler` has no syntactic counterpart, I'd say we simply replace its occurrences with `Block`.  What do you think?

Reply:
At the end of my review I convinced myself that this page should map _from_ C/C++ syntax _to_ QL classes. That means `Handler` and `CatchBlock` belong in the same row, and the syntax example should be the one that's currently on `CatchBlock`.

Reply:
The syntax example for `CatchBlock` includes a reference to `Handler`, so adding `Handler` in the QL class column would lead to`Handler` being defined in terms of itself.  I think what I will do is change the syntax example to refer to `Block` instead of `Handler`.

Reply:
Remember, `Handler` is the element _above_ each catch block, not below. Also, a `CatchBlock` doesn't _contain_ a `Block` -- it _is_ a `Block`. Putting it all together, here's the table row I suggest:

| C/C++ syntax | QL class(es) |
| ----- | ----- |
| `catch(Type parameter) { … }` | `CatchBlock`, `Handler` (as links) |
| `catch(...) { … }` | `CatchAnyBlock`, `Handler` (as links) |